### ISSUE TITLE
[3.3.5] Corpse visibility

### ISSUE THREAD

**Issue by Trisjdc** [TrinityCore/TrinityCore#16685]
https://github.com/TrinityCore/TrinityCore/commit/a3b2144750d4824a6b3d94766e7019f940a9a470

When a corpse first spawns, you can see the corpse, whether it be bones or not.
However... if you go beyond object visibility distance, and the corpse stops being visible, if you go near once again you won't be able to see the corpse.
Even more interesting is that the corpse is sent to the client, and the object update packets seem identical, whether the corpse is created near you or you get close enough to see it.

You can see the corpse again if you .tele near it. I don't know why this happens either, you'll see the corpse even if the object update packets were already sent and you're already near it.

Quick way to repeat the problem:
- .tele borean
- .die
- Release corpse
- .revive
- Walk to where you released, notice how you can't see the corpse
- .tele borean (which is where you are already)
- Now you can see corpse!

From what I could figure, it seems you can always see the corpse just fine if it's the *first time* the server sends a corpse with that GUID.
Maybe corpses aren't meant to ever be removed, client-side?

**Comment by r00ty-tc** [comment]
I vaguely remember @Shauren asking about this when implementing the map based corpse changes. E.g. it was already noticed this happens. Making the exact same observation, that corpse data is sent to client but it won't show it.

To me, the question is whether this also happens in TC 6.x, if so, we can maybe see what happens on retail in this situation?

**Comment by w5860363** [comment]
@r00ty-tc but Still where, GM is the only visible

**Comment by Killyana** [comment]
Still happen in  rev. 69c15f5ae769

**Comment by xvwyh** [comment]
I've tried multiple times to figure out why this bug is happening, but didn't succeed. I did, however, discover that it isn't as severe as I thought it was.

TL;DR: Corpses only stay invisible for 2 minutes and are repaired once they reenter the player's vision. Also this is 100% a client bug that can't be gracefully fixed from the server.

Longer explanation:

In order to optimize the client's performance, Blizzard's ObjectMgrClient doesn't immediately delete every object once it leaves the client's vision (or, rather, when the server reports it as being out of range), but instead keeps the object allocated in memory, in a "disabled" state, for 2 minutes, after which it's flagged for full deletion. If before these 2 minutes expire the object with the same GUID reappears in the vision - the old object is "reenabled" and adjusted according to the new data sent by the server (position, updatefields etc). If the 2 minutes have expired and the object was flagged for deletion - the next time the client receives a `SMSG_UPDATE_OBJECT` packet with ANY "out of range" GUIDs - it runs a routine that fully deletes all the clientside objects that were flagged for deletion.

For whatever reason, corpses in 3.3.5 break during that "reenable" procedure. I don't know the exact cause, my guess is that it has something to do with bounding boxes not being updated - corpses REAPPEAR if you have the map's 0,0,0 coordinates visible on screen (best place to test this - .tele test, that map has a floor right next to 0,0,0), but they have broken lighting. Corpses also get completely fixed if the player near-teleports somewhere, e.g. uses Blink - that causes the client to execute some code on every visible object and something inside that code fixes bugged corpses.

But all in all, if the players walks away from corpses, waits 2 minutes, and returns back - corpses will become visible again, because old corpses would be deleted from the client's ObjectMgrClient and brand new ones would be recreated, without the bug (yet). So this issue is not entirely permanent, and it kinda fixes itself over time, unless the player lingers in one area.

And to be super precise on this: at least one other object must disappear from the client's vision after 2 minutes have passed with corpses out-of-vision to trigger the cleanup procedure. Only after that the corpses will start working again when reentering vision.

There are probably some insanely overengineered ways to make a workaround for this client bug, like send corpses to the client with a new GUID each time they enter the client's vision, but that will require keeping track of which GUID was used for which player, in order to allow corpses to be deleted when exiting vision. I'd say it's not worth the effort.

---

Now for some other issues that **may or may** not be present on TC, I don't know, I just want to write how I dealt with those on my - very outdated - server:

### Players can't see their own corpse when they return to it after releasing spirit

This can happen if the player teleports to the same map when releasing the spirit, and they reach the corpse sooner than the 2 minute timer expires.  The reason for this is most likely that the client is informed about the corpse spawning (with `SMSG_UPDATE_OBJECT`) BEFORE the player gets teleported away. This causes the bug explained above to occur: client sees the corpse, then the corpse exits vision (when the teleport happens), then, as their ghost approaches the corpse - the corpse reenters vision, old object that wasn't deleted from memory yet is reenabled, and the corpse's visibility bugs out.

* **Solution:** Dunno, I only realized that this problem exists as I was writing this text. If I come up with a graceful fix for this - I'll update this comment.

   This issue arrived when objects started calling UpdateObjectVisibility immediately upon being added to the map. The ideal solution would be to somehow ensure that the newly-spawned corpse doesn't get sent to the player that's going to be teleported away. However at that stage (`Player::BuildPlayerRepop`) we don't know if the player will or will not be teleported away (`Player::RepopAtGraveyard`). Swapping the order of operation around isn't a solution either - that would create complications in trying to determine at which coordinates to spawn the corpse. This will also not account for the cases when the player PASSES BY their corpse as they're running to it, but then running out of vision range again.

   A hacky solution would be to NEVER remove the player's own corpse from vision (`Object::IsAlwaysVisibleFor` / `Object::CanAlwaysSee`), even when out of range - that shouldn't be difficult, because, thankfully, corpses don't get changed at runtime, so we don't need to worry about objectupdatenotifiers reaching the player.

### Bones don't show up after the player reclaims their own corpse

This can be important on BGs and BFs, as corpses/bones can drop loot for players.

This issues happens because when the player reclaims their corpse - the corpse is removed and bones are spawned as a brand new object BUT WITH THE SAME GUID AS THE CORPSE. This causes the exact same problem as the above: corpse is removed from vision, then the "very same" corpse (as far as the client is concerned - even tho bones are technically a new object on the server, because it reuses the same guid as the old corpse - to client it seems as if it's the old corpse is reentering vision) but now with `CORPSE_FLAG_BONES` enters the vision again, object reenable routine breaks it, and bones never become properly visible.

* **Solution 1 (hack):** Generate a new GUID for the bones, don't reuse the corpse's GUID.
* **Solution 2 (blizzlike):** Don't create a new object for bones, CHANGE the old corpse to bones. This one has a trick to it:
   Non-bone corpses contain updatefields that represent the equipment the player was wearing when they died. Bones don't need that data anymore, so it may as well be zeroed-out when converting a corpse to bones. However, that will cause extra unnecessary data to be sent in `SMSG_UPDATE_OBJECT` with 19 fields being changed to 0. All you need to do is set `CORPSE_FLAG_BONES`, the client will do all the other cleanups for you. You can then change the values of all the other updatefields to 0 SILENTLY, without triggering updates to be sent to clients. (code below, it breaks markdown formatting)

   This is exactly how blizzard does it. If you examine sniffs, you'll see that no updates are sent about `CORPSE_FIELD_ITEM`s changing to 0, yet the next time the same bones reenter the vision - no `CORPSE_FIELD_ITEM`s fields are sent either, because they have actually been changed to 0 serverside, and 0 fields aren't normally reported.

   There is one other caveat with this solution: non-bone corpses are worldobjects, bones - aren't. If you convert corpses to bones - they will remain worldobjects, downgrading performance. If you go this route - you also need to add support for switching corpses between worldobject states (see `Object::m_isTempWorldObject`, `Map::AddObjectToSwitchList` and  `Map::RemoveAllObjectsInRemoveList()`). This may also come in handy in the future, should you decide to implement corpses attaching to transports.

```cpp
void Corpse::ConvertToBones()
{
    m_type = CORPSE_BONES;
    ResetGhostTime();
    RemoveFlag(CORPSE_FIELD_FLAGS, CORPSE_FLAG_LOOTABLE);
    SetFlag(CORPSE_FIELD_FLAGS, CORPSE_FLAG_BONES);
    SetWorldObject(false);

    // Just erase items, don't send updates to players. This is blizzlike
    for (uint8 i = 0; i < EQUIPMENT_SLOT_END; ++i)
        m_uint32Values[CORPSE_FIELD_ITEM + i] = 0;
}
```

### Bonus bug: players can't loot player corpses/bones after relogging

Corpses store their loot recipient by pointer, not by GUID (`Corpse::lootRecipient`). If the player that qualified for the loot logs out and back in - they will no longer be able to loot the corpse/bones, because their `Player` has most likely been allocated at another memory address.

Thankfully, the pointer is never dereferenced, only compared for equality, so you don't get a crash from that. Regrettably, because it never crashed - you never noticed this bug.

Duplicated in #28157.

### Bonus bug: dead players can see objects around the corpse's position even if they're on a different map

If the player dies, let's say, in a dungeon, and releases spirit - if they happen to walk on the continent map near the same XYZ coordinates as their corpse in the dungeon map - they will suddenly start seeing nearby creatures, gameobject and players wandering about.

The reason behind this is simple: there is no check whether the player and the corpse are in the same map:

https://github.com/TrinityCore/TrinityCore/blob/e11613eeb5dcad16f5908c9c7dd8109baabe9272/src/server/game/Entities/Object/Object.cpp#L1563-L1573
